qadence/ml_tools/callbacks/writer_registry.py

```diff
@@ -0,0 +1,375 @@
+        """
+        raise NotImplementedError("Writers must implement an open method.")
+
+    def close(self) -> None:
+        """Closes the writer and finalizes logging."""
+        raise NotImplementedError("Writers must implement a close method.")
+
+    def write(self, result: OptimizeResult) -> None:
+        """
+        Logs the results of the current iteration.
+
+        Args:
+            result (OptimizeResult): The optimization results to log.
+        """
+        raise NotImplementedError("Writers must implement a write method.")
+
+    def log_hyperparams(self, hyperparams: dict) -> None:
+        """
+        Logs hyperparameters.
+
+        Args:
+            hyperparams (dict): A dictionary of hyperparameters to log.
+        """
+        raise NotImplementedError("Writers must implement a log_hyperparams method.")
+
+    def plot(
+        self,
+        model: Module,
+        iteration: int,
+        plotting_functions: tuple[PlottingFunction, ...],
+    ) -> None:
+        """
+        Logs plots of the model using provided plotting functions.
+
+        Args:
+            model (Module): The model to plot.
+            iteration (int): The current iteration number.
+            plotting_functions (tuple[PlottingFunction, ...]): Functions used to
+                generate plots.
+        """
+        raise NotImplementedError("Writers must implement a plot method.")
+
+    def log_model(
+        self,
+        model: Module,
+        dataloader: Union[None, DataLoader, DictDataLoader],
+    ) -> None:
+        """
+        Logs the model and associated data.
+
+        Args:
+            model (Module): The model to log.
+            dataloader (DataLoader | DictDataLoader | None): DataLoader to use
+                for model input.
+        """
+        raise NotImplementedError("Writers must implement a log_model method.")
+
+    def print_metrics(self, result: OptimizeResult) -> None:
+        """Prints the metrics and loss in a readable format.
+
+        Args:
+            result (OptimizeResult): The optimization results to display.
+        """
+
+        # Find the key in result.metrics that contains "loss" (case-insensitive)
+        loss_key = next((k for k in result.metrics if "loss" in k.lower()), None)
+        if loss_key:
+            loss_value = result.metrics[loss_key]
+            msg = f"Iteration {result.iteration: >7} | {loss_key.title()}: {loss_value:.7f} -"
+        else:
+            msg = f"Iteration {result.iteration: >7} | Loss: None -"
+        msg += " ".join([f"{k}: {v:.7f}" for k, v in result.metrics.items() if k != loss_key])
+        print(msg)
+
+
+class TensorBoardWriter(BaseWriter):
+    """Writer for logging to TensorBoard.
+
+    Attributes:
+        writer (SummaryWriter): The TensorBoard SummaryWriter instance.
+    """
+
+    def __init__(self) -> None:
+        self.writer = None
+
+    def open(self, config: TrainConfig, iteration: int = None) -> SummaryWriter:
```

Comment:
I see there are a lot of `if self.writer` checks in the other methods, which require the caller to call `open` on the writer. Is there a reason why `open` shouldn't be executed directly in the `__init__`?

Reply:
We will need to have a method to set up the writer in case the log_folder changes during runtime. For example when create_subfolders_per_run = True, a new writer will have to be set up for each run (based on a new log_dir), or we will have to modify the existing writer - both of which will need a writer opening method at the start of training.